### DISCUSSION TITLE
How to train NER and Relation Extraction model together

### DISCUSSION THREAD

**Discussion by karndeepsingh** [explosion/spaCy#9139] [Help: Coding & Implementations] [answered]
Hi,
I want to train NER and Relation Extraction model together instead of training separately. Please help me to understand the flow from scratch. Following are my queries :
1. How to prepare dataset for NER and Relation. I am using Label Studio for annotation of NER+ RELATION but I don't know how should I convert it into spacy binary format. Here is the example output of annotation when I annotate it from label studio.

```
[
  {
    "id": 6,
    "annotations": [
      {
        "id": 3,
        "completed_by": {
          "id": 2,
          "email": "kdsinghsdl@gmail.com",
          "first_name": "",
          "last_name": ""
        },
        "result": [
          {
            "value": {
              "start": 9,
              "end": 63,
              "text": "Synergy One Lending, Inc. dba Mutual of Omaha Mortgage",
              "labels": [
                "PARTY NAME"
              ]
            },
            "id": "fkVqYYR_P7",
            "from_name": "label",
            "to_name": "text",
            "type": "labels"
          },
          {
            "value": {
              "start": 0,
              "end": 8,
              "text": "BORROWER",
              "labels": [
                "PARTY ROLE"
              ]
            },
            "id": "62m6jvJopr",
            "from_name": "label",
            "to_name": "text",
            "type": "labels"
          },
          {
            "value": {
              "start": 64,
              "end": 102,
              "text": "5716 Corsa Avenuew, Sulle 102 Westlake",
              "labels": [
                "PARTY ADDRESS"
              ]
            },
            "id": "ZMBV98QR7N",
            "from_name": "label",
            "to_name": "text",
            "type": "labels"
          },
          {
            "from_id": "62m6jvJopr",
            "to_id": "fkVqYYR_P7",
            "type": "relation",
            "direction": "right",
            "labels": [
              "ROLE"
            ]
          },
          {
            "from_id": "ZMBV98QR7N",
            "to_id": "fkVqYYR_P7",
            "type": "relation",
            "direction": "right",
            "labels": [
              "ADDRESS"
            ]
          }
        ],
        "was_cancelled": false,
        "ground_truth": false,
        "created_at": "2021-09-04T08:14:02.157201Z",
        "updated_at": "2021-09-04T08:14:02.157201Z",
        "lead_time": 80162.164,
        "prediction": {},
        "result_count": 0,
        "task": 6
      }
    ],
    "predictions": [],
    "file_upload": "New_Text_Document_2_E6Don9E.txt",
    "data": {
      "text": "BORROWER Synergy One Lending, Inc. dba Mutual of Omaha Mortgage 5716 Corsa Avenuew, Sulle 102 Westlake village."
    },
    "meta": {},
    "created_at": "2021-09-03T09:54:27.365638Z",
    "updated_at": "2021-09-03T09:54:27.365638Z",
    "project": 6
  }
]
```

Above is the output file when I annotate for Relation. Please guide me what are the key value pairs required to make it in spacy binary format.

2. Do I also need to save seperate annotated dataset for NER. As I want to club NER + RELATION, so, do I need to prepare separate annotated Dataset for NER or this relation annotated dataset will also work.

3. After annotated dataset is convert to spacy binary format what changes shall i do in files to get these NER + RELATION trained together?

Please help me to understand the flow. It will be a great help.
Thanks

**Comment by polm** [comment, marked as answer]
Don't use all capitals when creating an issue.

As mentioned in the [training data section of the docs](https://spacy.io/usage/training#training-data), training data in spaCy just consists of making the `Doc` objects you want to see as output from your model. One thing that's a little tricky is that you can't specify different training data for different components - if you want to train multiple components at the same time, you need give the same data to all of them. In this case this means that your training data needs both relation annotations and NER annotations on each individual doc. Because training data for relation extraction already includes entity labels you should just be able to use your relation extraction training data as is for NER too.

I think you've already trained the components separately, but the NER annotations go in `doc.ents` and the relations go in `doc._.rel`. See the [relation extraction sample project for an example](https://github.com/explosion/projects/tree/v3/tutorials/rel_component). 

For some components you can provide special annotations that are to just be ignored, usually by giving a "-" label. This works for NER but I don't think there's anything equivalent for relation extraction. This means you should annotate all your data with both NER and relations. 

For the config, you just need to specify NER and relation extraction sections. The components should be in the pipeline in that order. You can see what a pipeline with multiple components looks like by using the Quickstart and specifying NER and textcat, for example. You might also want to experiment with specifying NER as an [annotating component](https://spacy.io/usage/training#annotating-components), since relation extraction uses the entity annotations.

**Comment by karndeepsingh** [comment]
Hi @polm ,
Thanks for detailed answer. The structure of annotation output that I have shared after labelling a single text from label studio looks good to you? If this output structure is fine and now additional changes are required then, I can directly convert this annotation json file to .spacy binary file?

I am unable to understand whether this annotation structure is fine or it need changes before I convert this json annotated output to .spacy binary format.

**Reply by polm** [reply]
Your JSON does not look like spaCy JSON, so you can't just use `spacy convert`. You need to make a `Doc` and set the annotations. See [this script](https://github.com/explosion/projects/blob/v3/tutorials/rel_component/scripts/parse_data.py) for an example for relation extraction.

**Reply by karndeepsingh** [reply]
@polm Okay! Can you share Spacy json structure? If I could write my own script and convert my annotation into the required spacy json format and structure, I guess after this I may be able to use that script you mentioned.

Please also suggest some annotation tool which I can use for NER + Relation.

Thanks

**Reply by polm** [reply]
The [JSON training format](https://spacy.io/api/data-formats#json-input) is deprecated. It's kind of complicated to make, and if you don't already have it I wouldn't bother making it.

Instead you just need to write a little code to set your annotations on Docs, similar to the script I linked or the training data section in the docs.

**Reply by karndeepsingh** [reply]
@polm you referring to this [script](https://github.com/explosion/projects/blob/v3/tutorials/rel_component/scripts/parse_data.py) right ?

**Reply by polm** [reply]
Yes, that's the one.

**Reply by karndeepsingh** [reply]
Okay! I will write a simple code to do it. And once I have annotation, then I need to add "ner" and "relation extractor" in pipeline in config file and mention the train.spacy, test.spacy and dev.spacy and start the training using spacy train all_gpu command!

Correct me, If my flow is wrong!

**Reply by polm** [reply]
If you have train, dev, and test files, you don't provide the test file to spaCy, just the first two.

If you're using the sample project as a base then `all_gpu` should be fine, though it's separate from `spacy train` - you would use it with `spacy project run all_gpu`. You also don't have to use a GPU.

**Reply by karndeepsingh** [reply]
Great! I will try it! First objective is to prepare the Json structure in required format and convert it into spacy binary format. 

Can you also recommend any good annotation tool that would ease up this process?

**Reply by polm** [reply]
I don't think there's anything wrong with your annotations in their current format, but Prodigy is a great tool that we make and it has a [combined NER/relations annotation mode](https://prodi.gy/docs/dependencies-relations#ner).

As I mentioned before, you don't have to make JSON in any particular format. You just need to make docs with annotations like in the script I linked.

**Reply by karndeepsingh** [reply]
Sure! I will edit that script according my annotated data and prepare the binary file!

Thanks. I will let you know once I am able to do this!

**Reply by karndeepsingh** [reply]
@polm  I don't have "token_start" and "token_end" in my annotation (shared at the top). What changes I need to add according to my annotations in parser_data.py file?
![image](https://user-images.githubusercontent.com/49562460/132136386-c8e78a7e-b1ed-42f4-934a-ba65a86a6414.png)

**Reply by karndeepsingh** [reply]
> @polm you referring to this [script](https://github.com/explosion/projects/blob/v3/tutorials/rel_component/scripts/parse_data.py) right ?

Is this script also prepares the data for NER ? or it is just for relation? Because I want to train NER + Relation, how both annotations can be added ? Does this script already consider NER + Relation annotation? If not, then what shall I add in order to get NER + RELATION binary file generated?

**Reply by karndeepsingh** [reply]
> @polm I have prepared the .spacy binary file for single sample. And when I pass it for training relations, it throws error, I am not able to understand whether .spacy binary file the information it carries is not correct or I am passing single sample as train , test for just checking, it is creating a problem?
> 
> ================================= train_gpu =================================
> Running command: /usr/bin/python3 -m spacy train configs/rel_trf.cfg --output training --paths.train data/relations_training.spacy --paths.dev data/relations_dev.spacy -c ./scripts/custom_functions.py --gpu-id 0
> ℹ Saving to output directory: training
> ℹ Using GPU: 0
> 
> =========================== Initializing pipeline ===========================
> [2021-09-06 10:04:53,186] [INFO] Set up nlp object from config
> [2021-09-06 10:04:53,201] [INFO] Pipeline: ['transformer', 'relation_extractor']
> [2021-09-06 10:04:53,211] [INFO] Created vocabulary
> [2021-09-06 10:04:53,212] [INFO] Finished initializing nlp object
> Some weights of the model checkpoint at roberta-base were not used when initializing RobertaModel: ['lm_head.layer_norm.bias', 'lm_head.bias', 'lm_head.dense.bias', 'lm_head.layer_norm.weight', 'lm_head.decoder.weight', 'lm_head.dense.weight']
> 
> * This IS expected if you are initializing RobertaModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
> * This IS NOT expected if you are initializing RobertaModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
>   [2021-09-06 10:04:57,689] [INFO] Initialized pipeline components: ['transformer', 'relation_extractor']
>   ✔ Initialized pipeline
> 
> ============================= Training pipeline =============================
> ℹ Pipeline: ['transformer', 'relation_extractor']
> ℹ Initial learn rate: 0.0
> E # LOSS TRANS... LOSS RELAT... REL_MICRO_P REL_MICRO_R REL_MICRO_F SCORE
> 
> ⚠ Aborting and saving the final best model. Encountered exception:
> KeyError("[E900] Could not run the full pipeline for evaluation. If you
> specified frozen components, make sure they were already initialized and
> trained. Full pipeline: ['transformer', 'relation_extractor']")
> Traceback (most recent call last):
> File "/usr/local/lib/python3.7/dist-packages/spacy/training/loop.py", line 281, in evaluate
> scores = nlp.evaluate(dev_corpus(nlp))
> File "/usr/local/lib/python3.7/dist-packages/spacy/language.py", line 1389, in evaluate
> results = scorer.score(examples)
> File "/usr/local/lib/python3.7/dist-packages/spacy/scorer.py", line 135, in score
> scores.update(component.score(examples, **self.cfg))
> File "/content/rel_component/scripts/rel_pipe.py", line 201, in score
> return score_relations(examples, self.threshold)
> File "/content/rel_component/scripts/rel_pipe.py", line 211, in score_relations
> gold_labels = [k for (k, v) in gold[key].items() if v == 1.0]
> KeyError: (0, 4)
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
> File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
> "**main**", mod_spec)
> File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
> exec(code, run_globals)
> File "/usr/local/lib/python3.7/dist-packages/spacy/**main**.py", line 4, in
> setup_cli()
> File "/usr/local/lib/python3.7/dist-packages/spacy/cli/_util.py", line 69, in setup_cli
> command(prog_name=COMMAND)
> File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 829, in **call**
> return self.main(*args, **kwargs)
> File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 782, in main
> rv = self.invoke(ctx)
> File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 1259, in invoke
> return _process_result(sub_ctx.command.invoke(sub_ctx))
> File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 1066, in invoke
> return ctx.invoke(self.callback, **ctx.params)
> File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 610, in invoke
> return callback(*args, **kwargs)
> File "/usr/local/lib/python3.7/dist-packages/typer/main.py", line 497, in wrapper
> return callback(**use_params) # type: ignore
> File "/usr/local/lib/python3.7/dist-packages/spacy/cli/train.py", line 63, in train_cli
> train(nlp, output_path, use_gpu=use_gpu, stdout=sys.stdout, stderr=sys.stderr)
> File "/usr/local/lib/python3.7/dist-packages/spacy/training/loop.py", line 122, in train
> raise e
> File "/usr/local/lib/python3.7/dist-packages/spacy/training/loop.py", line 105, in train
> for batch, info, is_best_checkpoint in training_step_iterator:
> File "/usr/local/lib/python3.7/dist-packages/spacy/training/loop.py", line 226, in train_while_improving
> score, other_scores = evaluate()
> File "/usr/local/lib/python3.7/dist-packages/spacy/training/loop.py", line 283, in evaluate
> raise KeyError(Errors.E900.format(pipeline=nlp.pipe_names)) from e
> KeyError: "[E900] Could not run the full pipeline for evaluation. If you specified frozen components, make sure they were already initialized and trained. Full pipeline: ['transformer', 'relation_extractor']"
> 
> ✘ Missing dependency specified by command 'evaluate':
> training/model-best
> Maybe you forgot to run the 'project assets' command or a previous step?

**Reply by karndeepsingh** [reply]
@polm Hey, I have fixed all issues and train ner + relation in a single pipeline. I have a question.. How to do inferencing of both ner and relation models together? Any sample code would help.Thanks

**Reply by MaticBernik** [reply]
You figured out what was wrong with your input documents to give you this error?

> File "/content/rel_component/scripts/rel_pipe.py", line 211, in score_relations
> gold_labels = [k for (k, v) in gold[key].items() if v == 1.0]
> KeyError: (0, 4)

**Reply by Rithik-Alias** [reply]
Hey @karndeepsingh, did you get the error solved? I am also facing the same issue.

**Reply by polm** [reply]
@Rithik-Alias There are a lot of different things happening in this thread, so it's not clear what you're referring to. It would be better for you open a new Discussion and clearly state what error you're seeing and what caused it.

**Reply by Rithik-Alias** [reply]
@polm Thank you for replying to it. I have started a new discussion : [https://github.com/explosion/spaCy/discussions/9551](https://github.com/explosion/spaCy/discussions/9551)

**Reply by panchalsp** [reply]
> > @polm you referring to this [script](https://github.com/explosion/projects/blob/v3/tutorials/rel_component/scripts/parse_data.py) right ?
> 
> Is this script also prepares the data for NER ? or it is just for relation? Because I want to train NER + Relation, how both annotations can be added ? Does this script already consider NER + Relation annotation? If not, then what shall I add in order to get NER + RELATION binary file generated?

Hey @karndeepsingh @polm Did this [script](https://github.com/explosion/projects/blob/v3/tutorials/rel_component/scripts/parse_data.py) work for both NER + Relation?

**Reply by Faizan-Siddique** [reply]
Hi @karndeepsingh if possible could you share your script for conversion f label studio annotation json file into spacy binary format thanx in advance

**Comment by karndeepsingh** [comment]
@polm  How I can visualize the relation and entities using DisplayC ? I can only see NER is being displayed by using Displayc but how both relation and Entities can be displayed together?

Thanks

**Reply by polm** [reply]
displaCy doesn't have direct support for displaying relations, you'd need to customize the output.

**Reply by karndeepsingh** [reply]
Okay thanks @polm

**Reply by Lolologist** [reply]
@karndeepsingh there are other tools out there that can, though! I think Label Studio is one.

**Reply by karndeepsingh** [reply]
@Lolologist @polm Hi, I am using Label Studio to do the annotation. But it doesn't output "token_start" and "token_end" when I annotate a document as token_start and token_end output is given in prodigy. So, I have to calculate this "token_Start" and "token_end" through code while creating .spacy binary file. Can you please help me to understand how I can achieve "token_start" and "token_end" for every document like what is the logic to get token_start or token_end. 

Thanks

**Reply by polm** [reply]
I asked you to not @ people before and Sofie just told you not to do it in another discussion. Stop doing that.

**Reply by Lolologist** [reply]
Karneepsingh, you will likely need to make your own connector that translates between LS and spaCy. I'm afraid that's likely outside the scope of this thread and explosion's interests, however.

**Comment by Lance-Owen** [comment]
@karndeepsingh Hello, friend, I also want to make a ner and relation together model, but I don't know how to configure config, please ask, can you share your method, or config file, if you can thank you very much!
